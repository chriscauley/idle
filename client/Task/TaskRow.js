import { uniq } from 'lodash'
import React from 'react'
import css from '@unrest/css'
import { post, Dropdown, SchemaForm } from '@unrest/core'

import api from '../api'
import Task from './model'
import { makeTypeahead } from './Typeahead'

const getLinks = ({ task, activity, edit }) => {
  const links = [{ children: 'Edit Task', onClick: edit }]
  if (activity) {
    links.push({
      children: 'Edit Activity',
      to: `/activity/${activity.id}/edit/`,
    })
    links.push({
      children: 'View Activity',
      to: `/activity/${activity.id}/`,
    })
  } else {
    links.push({
      children: 'Create Activity',
      to: `/activity/${task.id}/from_task/`,
    })
  }

  return links
}

const fields = {
  DateTime: (value) => ({
    type: 'string',
    format: 'date-time',
    default: value ? new Date(value).toString() : undefined,
  }),
}

const getTextsOptions = (key) => () => {
  const { tasks = [] } = api.task.use()
  return uniq(tasks.reduce((acc, item) => acc.concat(item[key] || []), [])).filter(Boolean)
}

function ActiveTaskForm({ task, activity, editing, setEditing }) {
  const { setData, tasks } = api.task.use()
  const onChange = (data) => {
    // update the task in the /api/task/ rest list so the TaskRow matches the form.
    // task is already in tasks, so we just re-setData the tasks to update tasks in list.
    if (Object.keys(data).find((key) => data[key] !== task[key])) {
      Object.assign(task, data)
      setData({ tasks })
    }
  }
  const properties = {}
  const required = []
  const uiSchema = {}
  if (task.completed) {
    properties.started = fields.DateTime(task.started)
    properties.completed = fields.DateTime(task.completed)
  }
  if (!task.started) {
    properties.due = fields.DateTime(task.due)
  }
  activity.boolean_fields?.forEach((s) => {
    properties[s] = {
      type: 'boolean',
      default: task[s],
    }
  })
  activity.measurements?.forEach((s) => {
    required.push(s)
    properties[s] = {
      type: 'number',
      default: task[s],
    }
  })
  activity.texts?.forEach((s) => {
    required.push(s)
    properties[s] = {
      type: 'string',
      default: task[s],
    }
    uiSchema[s] = {
      'ui:field': makeTypeahead(getTextsOptions(s)),
    }
  })
  if (Object.keys(properties).length === 0) {
    return null
  }
  const schema = { type: 'object', properties, required }
  return (
    <div className="flex-grow w-full">
      <SchemaForm
        prepSchema={() => schema}
        onChange={onChange}
        uiSchema={uiSchema}
        form_name={`ActiveTaskForm/${task.id}`}
        autosubmit={true}
        customButton={true}
      >
        {editing && (
          <button className={css.button()} onClick={() => setEditing(false)}>
            Done
          </button>
        )}
      </SchemaForm>
    </div>
  )
}

const trigger = ({ _activity, task, refetch }) => {
  const now = new Date().valueOf()
  const data = {}
  if (task.completed) {
    data.completed = null
  } else if (task.started) {
    data.completed = now
  } else {
    data.started = now
  }
  post(`/api/task/${task.id}/`, data).then(() => refetch())
}

export default function TaskRow({ task, editing, setEditing }) {
  const { refetch, tasks } = api.task.use()
  const { activities } = api.activity.use()
  if (!(tasks && activities)) {
    return null
  }
  const activity = activities.find((a) => a.id === task.activity_id)
  const showForm = editing || (task.started && !task.completed)
  const edit = () => setEditing(task.id)

  return (
    <li className={css.list.item('flex-wrap')} key={task.id}>
      <div className="flex w-full items-center">
        <div className="trigger">
          <i
            className={css.icon(Task.getIcon(task))}
            onClick={() => trigger({ activity, task, refetch })}
          />
        </div>
        <span className="flex-grow">
          <div>{task.name}</div>
          <div className="flex justify-between w-full">
            <div className="light">{Task.getTime(task)}</div>
            <div className="light">{Task.getShortMeasures({ task, activity })}</div>
          </div>
        </span>
        <Dropdown
          links={getLinks({ task, activity, edit, editing })}
          title={<i className={css.icon('ellipsis-v')} />}
        />
      </div>
      {showForm && <ActiveTaskForm {...{ task, activity, editing, setEditing }} />}
    </li>
  )
}
