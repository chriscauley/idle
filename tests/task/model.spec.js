import Task from '../../client/task/model'

const now = new Date().valueOf()
const HOUR = 1000 * 3600
const DAY = HOUR * 24

// in all these cases, completed then started then due should have priority and the others ignored
const timeTasks = [
  { name: 'overdue', due: now - HOUR },
  { name: 'due soon', due: now + HOUR },
  { name: 'started', started: now - HOUR, due: now - HOUR },
  {
    name: 'completed',
    completed: now - DAY,
    due: now - 2 * HOUR,
    started: now - HOUR,
  },
]

test('getIcon', () => expect(timeTasks.map(Task.getIcon)).toMatchSnapshot())
test('getTime', () => expect(timeTasks.map(Task.getTime)).toMatchSnapshot())
test('getShortMeasure', () => {
  const activity = { measurements: ['milk', 'eggs', 'cheese'] }
  const task = { milk: 5, eggs: 0 }
  expect(Task.getShortMeasures({ activity, task })).toMatchSnapshot()
})

test('sortByDate', () => {
  const permArr = []
  const usedChars = []

  function permute(input) {
    let i, ch
    for (i = 0; i < input.length; i++) {
      ch = input.splice(i, 1)[0]
      usedChars.push(ch)
      if (input.length === 0) {
        permArr.push(usedChars.slice())
      }
      permute(input)
      input.splice(i, 0, ch)
      usedChars.pop()
    }
    return permArr
  }
  const [ans, ...rest] = permute([...timeTasks]).map((tasks) =>
    Task.sortByDate(tasks)
      .map((t) => t.name)
      .join('|'),
  )
  expect(ans).toMatchSnapshot()
  expect(rest.length).toEqual(23)
  rest.forEach((s) => expect(s).toEqual(ans))
})
