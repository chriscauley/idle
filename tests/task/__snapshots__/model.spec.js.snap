// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getIcon 1`] = `
Array [
  "square-o",
  "square-o",
  "spinner fa-spin fa-pulse",
  "check-square-o",
]
`;

exports[`getShortMeasure 1`] = `"m5 e0 c?"`;

exports[`getTime 1`] = `
Array [
  "1 hour overdue",
  "due in 1 hour",
  "started 1 hour ago",
  "done 1 day ago",
]
`;

exports[`sortByDate 1`] = `"started|due soon|overdue|completed"`;
